### ISSUE TITLE
Open web page or copy URL/username to clipboard.

### ISSUE THREAD

**Issue by nikolay-turpitko** [gopasspw/gopassbridge#23]
I like this plugin's minimalist UI. But it's not quite convenient to open page with it.

I'm in the process of migrating my passwords from LastPass, and before LastPass I used Clipperz (I even implemented my own Chrome plugin for Clipperz, but abandoned it long ago in favor of LastPass).

I know `gopass` and `gopassbridge` uses yaml-like password entries format and `gopassbridge` leverage it to extract username. It'd be convenient to extract other custom fields as well (pin codes/security answers). Extracted values should be put into clipboard or opened in the viewer (external editor or right in the browser).

Both with LastPass and Clipperz I accustomed to use password manager as kind of replacement of bookmark manager to open web pages and fill login forms in one click. In case of `gopassbridge` it'd be convenient (and easy to implement) to copy URL/username into clipboard with some other keyboard modifier. `shift` is used to copy password. So, to copy URL and username either `ctrl` or `alt` or their combinations can be used. Action to open the page would be mouse click without modifier keys on empty tab (currently this action used to fill the login form on loaded page, this code could be modified to check if domain name is empty and to fill it from clicked item).

Ability to open web page and auto-fill it's login form via plugin would be a nice feature, but in case if it hard to implement or won't work for some reason, ability to copy URL/username to clipboard would be a good fallback.

Another my requirement to password manager is to quickly view extra notes, stored with password. Such extra notes may contain some information, required during login. Like pin codes or security answers. Even with LastPass it wasn't very smooth process to access such web sites - I had to find password entry in the plugin, click on it and when it opens and fills login form, I had to find entry again and open additional info in LastPass' web editor. In case of `gopassbridge` it's even more complicated and requires to switch to terminal to request `gopass` for extra info. I wonder is it possible to implement necessary functionality in the plugin, keeping it's simplistic UI. Fore example, it'd be some plugin setting to automatically execute some CLI command on such pages. It'd be very flexible. For example, I'd add yaml tag like `execute-cmd: notifysend "xxx"` or `execute-cmd: "x-terminal-emulator gopass ..."`. So that it notified me or opened entry in terminal window when I open particular page. User would be able to easily customize command for his own needs and OS. It's pure speculation, probably it's not the best approach. Another option is to open web page with custom fields automatically if there are extra tags and some flag is set in plugin's options.

Yet another idea. Would it be nice to configure "username" tag name via plugin options and optionally ignore the case? Currently they are hard-coded and, for example, camel case or upper case is not recognized.

Besides using key modifiers it's possible to use right-mouse-click custom context menu for plugin and put some options there. For example, I'd add such options:

- open URL (and fill form automatically when it opened);
- copy URL to clipboard;
- copy username to clipboard;
- copy extra yaml tag to clipboard (if it's possible to list all yaml tags, then add them as separate items, otherwise just copy all of them at once);
- open password entry in a browser page (like `gopass show ... | tail -n +3`);
- open password entry with external command (like cut to new terminal window or edit in default editor).

It's a huge blob of raw ideas. I'm not sure what option would be simpler to use/implement and won't unnecessarily complicate things. Let's discuss, probably extract some simpler tasks from it and I'll try to help with implementation in my spare time (which I have not much, I should admit).

**Comment by nikolay-turpitko** [comment]
@martinhoefling

I see that actually you don't have URL in the jsonapi and gopass cannot extract it. It's my own convention to save URL as yaml field within password entry. Hardcoding this convention will impose it to all users, while pass allows to be flexible with storage format. From the other side, you already force users to use "login/username" field and yaml structure...

What if add a request to `gopass jsonapi` to get whole password entry, parsed as yaml and converted to json? Probably, plus full entry path and password as special fields. Something like:

```
{
 "_path": "my/entry/path/www.domain.com/something",
 "_pass": "xxx",
 "urls": [ "http://www.domain.com", "http://sso.login-for-domain.com" ],
 "login-name-in-fancy-format": "joe",
 "whatever-extra-field": "...",
 "_extra": "all the text not parsed as yaml"
} 
```

Is there some good JS templating library, allowing to make HTML page from it? Or just use Go templates to produce HTML and return it as json field. Then just render it right in extension's pop-up window? Template can be put somewhere near wrapper script and allowed to be amended by user. Some simple one would be installed with the gopass/gopassbridge.

This way user will be able to use suggested format from the box, or change template according to his needs. Template will produce HTML right within plugin's popup panel and it can contain hyperlinks to open URLs, stored in the password's entry, display some extra information. Don't know how, but, probably, it also will be able to execute some plugin's javascript function to pass it mapped username. Or this can be just setting of `gopass`, stored in it's config file. Something like `loginfield: User-Name`.

**Comment by martinhoefling** [comment]
Wow, that's a lot of ideas. So I also did consider already some of the ideas. One thing one would need to do first is to send all information from an entry with the jsonapi response. With this, i.e. the url stored in the `url` key can be entered. The same is true for other meta information. I actually thought about using an other modifier key (e.g.) alt+click on entry to show all data and from there make it possible to copy entries. The first step towards this would be to send send the information in the response.

**Comment by nikolay-turpitko** [comment]
Yeah, my first message is not quite focused, more like speculation about possibilities.
I think, it worth to extract something doable from it and create additional issues for less priority things.

I think that automatically filling a form, though cool feature, but require more work and tests and also not always work even in venerable LastPass. I found several important examples (online banks, payment systems) where for some (security?) reasons they show form fields one-by-one with necessity to press some button for next step. LastPass can deal with this, but even it has an option to fill fields manually. Also, I've seen some forms with field names, which LastPass could not guess and fill. So, I think, it's most important (and more easy to implement too) is to allow user to copy arbitrary piece of data from the password entry into arbitrary form field (or browser's address bar).

Another quite important feature is to recognise URLs within password entry and allow user to use them. For URLs, I think, most natural ways to use them is to open them in active tab. Though, I have some entries where I store URLs to pages with additional info or instructions. So, there can be situations, when user want to open page in the new tab.

So, I'd rename this issue and emphasise it's focus on showing/using arbitrary metadata from password entry and highlighting hyperlinks. 

Next, it's quite important to keep "current" search result for the active tab between plugin's popup window hides and re-opens. Currently plugin creates a list of items, matching host name of URL, opened in the current tab. It's cool, but doesn't work in situations, when host name is not good to be put into `gopass`'s entry path or metadata. For example, "Oracle Cert Center" is instantly recognisable name for the password entry, but it's login url can be something like "https://www9.pearsonvue.com/Dispatcher?v=W2L&application=Login&clientCode=ORACLE&...". Even worse, some web sites redirects for login to entirely different domain (single-sign-on, cannot quickly find real example). Point is if search result does not match current opened document it'll be closed. If I need to copy-paste several fields (username/password/pin), then I'll have to search it several times. BTW, LastPass is also no good at it. It also force me to navigate several times. What I'd suggest is to remember last one chosen (clicked) search item till the user switches the tab or clicks another search item. Such search item should be kept in the suggestion list above or below all other items. Regarding implementation I think it should be some map at background page, mapping opened documents/tabs onto items, clicked by user. Well, here is tricky part - when? User could navigate to the URL, clicking some hyperlink, displayed by the plugin. Then tab should be mapped onto password entry, used to open the tab. On the other hand, user could type url in the address bar, then click plugin's button, found that interesting item doesn't appear, and searched for item in the plugin's search field. Then he interacts somehow with result item (clicks it, opens context menu, copy one piece of data) at this point this result should be remembered. It's also hard to say at this point what would be keys and values of the map.

Another question is how to display metadata info. I'd just create context menu items for every yaml field to copy it to clipboard, except for hyperlinks, which makes sense to have option to open it. If context menu is difficult to create (I don't know API for it, but seems LastPass creates context menus for items), then another option is to "navigate" to generated page within popup. Page will contain "back" button to return to normal view and list of extracted items. They can appear like buttons or hyperlinks. For regular items action upon click would be copy it to clipboard (as for results now) and for links there should be couple of options - copy and open (probably in new vs current tab).

Regarding how many urls to extract from metadata and which one to use. I think it should be possible to extract all urls and allow user to decide which one to click. Moreover, I think, it's not necessary to require user to save urls in particular fields. URLs can be extracted with just a regexp (either on server's or on plugin's side). What if plugin just requested for all raw metadata in additional json request and display that metadata as text, only highlighting hyperlinks in it (and injecting actions to them). If it's possible to select text on popup page to copy it into clipboard, it'd be just enough. I quite often use several hyperlinks within password entries. Not all of them for the login form. Some are for additional instructions, etc. But I'd like all of them be clickable.

So, I think something like this, could be requested from the jsonapi:

```
{
 "path": "my/password/entry/path/like/gopath/see/it",
 "name": "xxx",
 "password": "xxx",
 "extractedURLs": [ "xxx", "xxx", ...],
 "extractedMetadata": {
   "whatever-keys": "xxx"
 },
 "raw": "all text, except password as is"
}

```

And than there are some flexibility how to render this data to user. For example, all "extractedMetadata" keys can be displayed in the list with option to copy them (either via click action, or just selecting by mouse in the popup view). All "extractedURLs" can be displayed as hyperlinks with action to open them, or open them in the new tab, etc.

WDYT?

**Comment by nikolay-turpitko** [comment]
![gopassbridge](https://user-images.githubusercontent.com/1192730/39109321-a1f85246-46f6-11e8-9bce-355af25fe69d.png)
I've tried to draw a sketch to illustrate this amount of text.

**Comment by nikolay-turpitko** [comment]
Having said that, I remembered another issue. Sometimes I use my notebook without mouse and with touchpad switched off. I use browser with shortcuts and vimium plugin, so it's possible to work with chrome like this. It'd be cool to be able to navigate plugin without mouse as well. With focus changing on TAB key. Form fields and hyperlinks will work OK by design. But it'd be difficult to select and copy arbitrary text without mouse. Whole line should accept keyboard focus and keypress action (similar to hyperlink) to make it possible. For example, whole field could be converted to hyperlink (but styled differently) in case it doesn't contain one. And field's name could be such a link too. For example, "Notes:" on the picture above may be clickable/keypressable to allow copying whole text, not only link within it (though, I'm not sure if it's necessary). "PIN" and "Phone" fields on the picture do not match hyperlink regexp, but it should be possible to navigate to them with TAB key and copy their values with ENTER.

**Comment by nikolay-turpitko** [comment]
Also, not shown on the sketch, but URLs and notes can be quite long. It's better to restrict plugin popup's size, but add scrollbars both to the list of search/suggestion results and to metadata box. And disallow lines of metadata to be wrapped. Probably, it even worth to implement some URL shortening technic (replace visible link to N chars with ...).

**Comment by martinhoefling** [comment]
Many good ideas and concepts. I try to summarize some key ideas / user stories and will open separate tickets / issues for them:

* As a user, if entry is selected on empty tab - load the content of the `url` key on select instead of trying to fill in password on an empty page. After load, try to fill in password (not sure if the latter is desired and possible) #24
* Keep the entered search until a new domain is loaded or tab is switched. If in detail view (see below), also keep selected detail view #25
* Detail view: If the user alt+clicks on an entry, show details similar as in the sketch above. Not sure if one wants to display the search results or just replace search with a detail view. #26

Other suggestions I am not sure about are:
* Opening execution commands (must be implemented in gopass, useful enough?)
* Right mouse click (to activate detail mode - possible on all browsers?)

Are you fine with that? Or is something missing?

**Comment by nikolay-turpitko** [comment]
> Opening execution commands (must be implemented in gopass, useful enough?)

I agree, it's better to remove it from the scope. I saw it as alternative way to display metadata. But with search details panel, containing all metadata, it's not that interesting.

> Right mouse click (to activate detail mode - possible on all browsers?)

Whatever. I'll be fine with modifier key or hyperlink/button on the popup page (probably, it's even better - more intuitive without need to read docs). Actually, as far as I can tell, Chrome/Chromium/Firefox/Opera/Safari on both Mac and Ubuntu/Debian can do it. LastPass plugin do it on those platforms. I'm pretty sure, it also possible on Windows, though I haven't used it for several years. I'd go for hyperlink. Actually, I'd replace default action when user clicked on search item with displaying it's details. And from details user can click on whatever hyperlink it likes. Another option, if you like it more, is to add small clickable icons to the and of the row.

> Are you fine with that? Or is something missing?

I'm fine with this, as long as there will be a (preferably quick) way to view and copy-paste different pieces of metadata. Cheers!

**Comment by martinhoefling** [comment]
closed by merging #24
